### PR TITLE
Retry reading .rrd files until end-of-stream message header

### PR DESCRIPTION
**Pull request by zehiko**
### What

* Fix for #4056

We introduce a retryable file reader that is used when reading `.rrd` files that are still being written and it simply ignored the EoF condition. We also introduce an *end-of-stream* message header that ensures we stop decoding (and endlessly trying to read the file) message stream once we see this header.

`.rrd` files generated with using a prior sdk version should work just fine, the only downside is that while the viewer is running there will be an open file handle for that file since no end-of-stream message will exist.

### Testing done
* viewer loading a file that's still being written 
* viewer loading a file with already written content
* streaming file from an http endpoint
* tests
```
cargo test --all-targets --all-features
```

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [ ] I've included a screenshot or gif (if applicable)
* [ ] I have tested the web demo (if applicable):
  * Using examples from latest `main` build: [rerun.io/viewer](https://rerun.io/viewer/pr/7475?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [rerun.io/viewer](https://rerun.io/viewer/pr/7475?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [ ] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG
* [ ] If applicable, add a new check to the [release checklist](https://github.com/rerun-io/rerun/blob/main/tests/python/release_checklist)!
* [ ] If have noted any breaking changes to the log API in `CHANGELOG.md` and the migration guide

- [PR Build Summary](https://build.rerun.io/pr/7475)
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

To run all checks from `main`, comment on the PR with `@rerun-bot full-check`.

### CODE REVIEW

**Review by emilk — COMMENTED** [review]
Looks good, but I'd like to have some tests for this.

First a unit-test to make sure reading concatenated files still work.

Then it would be nice to have a test that checks that the new retry-reader works, e.g. by actually writing some data to `/tmp` (https://crates.io/crates/tempfile), starting a decoder for it, and then appending more data to it

File: crates/store/re_data_loader/src/loader_rrd.rs [hunk 1]
```diff
@@ -36,22 +45,47 @@ impl crate::DataLoader for RrdLoader {
...
+                        let filepath = filepath.clone();
+                        move || {
+                            decode_and_stream(&filepath, &tx, decoder);
+                        }
+                    })
+                    .with_context(|| format!("Failed to open spawn IO thread for {filepath:?}"))?;
+            }
+            "rrd" => {
+                // For .rrd files we retry reading despite reaching EOF to support live (writer) streaming.
+                // Decoder will give up when it sees end of file marker (i.e. end-of-stream message header)
```
[the remark this answers is not in the document]
Comment (zehiko) [reply crates/store/re_data_loader/src/loader_rrd.rs:70]: I would also prefer to have a single path, but if I'm not mistaken `.rbl` files are loaded synchronously on the ui update path (I know I ended up making the viewer unresponsive initially :), so I thought there might be a risk  withreading blueprint files generated without the end-of-stream header.  Possibly I'm missing something obvious?

**Comment by zehiko** [comment]
> Looks good, but I'd like to have some tests for this.
> 
> First a unit-test to make sure reading concatenated files still work.
> 
> Then it would be nice to have a test that checks that the new retry-reader works, e.g. by actually writing some data to `/tmp` (https://crates.io/crates/tempfile), starting a decoder for it, and then appending more data to it

definitely agree, I was just looking for some initial feedback on the approach. Will improve the test coverage in the next revision (after I discuss the concatenation strategy offline).

File: crates/store/re_data_loader/src/loader_rrd.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (emilk) [reply crates/store/re_data_loader/src/loader_rrd.rs:70]: This sounds like a good reason yes, but then that should be explained in a comment :)

File: crates/store/re_log_encoding/src/decoder/stream.rs [hunk 2]
```diff
@@ -116,31 +116,42 @@ impl StreamDecoder {
...
-                        // parts of the app can easily access it.
-                        msg.info.store_version = self.version;
-                        Ok(Some(re_log_types::LogMsg::SetStoreInfo(msg)))
-                    } else {
-                        Ok(Some(message))
-                    };
+                    },
+                    MessageHeader::EndOfStream => {
+                        self.state = State::MessageHeader;
+                        return Ok(None);
```
[the remark this answers is not in the document]
Comment (zehiko) [reply crates/store/re_log_encoding/src/decoder/stream.rs:153]: as discussed offline, need to slightly change the retry strategy to always check if there's more data even if we see end-of-stream header.

File: crates/store/re_data_loader/src/loader_rrd.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (zehiko) [reply crates/store/re_data_loader/src/loader_rrd.rs:70]: agree, will address in the follow rev

File: crates/store/re_data_loader/Cargo.toml [hunk 3]
```diff
@@ -26,7 +26,7 @@ default = []
 [dependencies]
 re_build_info.workspace = true
 re_chunk.workspace = true
-re_log_encoding = { workspace = true, features = ["decoder"] }
+re_log_encoding = { workspace = true, features = ["decoder", "encoder"] }
```
Comment (emilk) [inline crates/store/re_data_loader/Cargo.toml:29]: Why does the data loader need an encoder?

File: crates/store/re_data_loader/Cargo.toml [hunk 4]
```diff
@@ -37,10 +37,11 @@ ahash.workspace = true
 anyhow.workspace = true
 arrow2.workspace = true
 image.workspace = true
-once_cell.workspace = true
 notify.workspace = true
+once_cell.workspace = true
 parking_lot.workspace = true
 rayon.workspace = true
+tempfile.workspace = true
```
Comment (emilk) [inline crates/store/re_data_loader/Cargo.toml:44]: If this is only used for test then it should be under `[dev-dependencies]`

**Review by emilk — APPROVED** [review]
Looks great!

Have you tested it manually, as a user would? You should be able to test it using `tests/rust/test_ui_wakeup` with:

```
cargo r -p test_ui_wakeup -- -- save stream.rrd
```

and then in another terminal `pixi run rerun stream.rrd`. Go back to the first terminal and press enter. A new message should be appended to the file, and the viewer should wake up and show it.

Feel free to add it as instructions to `tests/rust/test_ui_wakeup/src/main.rs` so we have a nice manual test for this that we can return to.

File: crates/store/re_data_loader/Cargo.toml (hunk 3 above)
Comment (emilk) [reply crates/store/re_data_loader/Cargo.toml:29]: Ah, if it is just for tests then add it to `[dev-dependencies]`

**Comment by zehiko** [comment]
> Looks great!
Have you tested it manually, as a user would?

Yes, with and without concatenation, using file that is actively being written to and also serving file from an http endpoint.

> You should be able to test it using tests/rust/test_ui_wakeup with:
cargo r -p test_ui_wakeup -- -- save stream.rrd
 and then in another terminal pixi run rerun stream.rrd. Go back to the first terminal and press enter. A new message should 
 be appended to the file, and the viewer should wake up and show it.
Feel free to add it as instructions to tests/rust/test_ui_wakeup/src/main.rs so we have a nice manual test for this that we can 
 return to.

Thanks for the pointers, tested using this as well, looks good, viewer is indeed updated on every 'enter'. Adding this new example to testing matrix as you suggested.

**Pull request merged**